pkg/execution/pauses/block.go

```diff
@@ -569,16 +593,222 @@ func (b *blockstore) PauseByID(ctx context.Context, index Index, pauseID uuid.UU
 
 // Compact reads all indexed deletes from block for an index, then compacts any blocks over a given threshold
 // by removing pauses and rewriting blocks.
-func (b *blockstore) Compact(ctx context.Context, idx Index) {
-	// Implement the following:
-	// TODO: Lease compaction for the index.
-	// TODO: Read all block metadata for the index
-	// TODO: Read all blockDeleteKey entries for the index
-	// TODO: For each deleted entry, record which block the delete is for.
-	// TODO: If len(block_deletes) > block_compact_ratio, rewrite the block by:
-	// 1. fetching the block
-	// 2. filtering deleted pauses from the block
-	// 3. rewriting the block
+func (b *blockstore) Compact(ctx context.Context, index Index) {
+	util.Lease(
+		ctx,
+		"compact index blocks",
+		// NOTE: Lease, Renew, and Revoke are closures because they need
+		// access to the Index field.  This makes util.Lease simple and
+		// minimal.
+		func(ctx context.Context) (ulid.ULID, error) {
+			return b.leaser.Lease(ctx, index)
+		},
+		func(ctx context.Context, leaseID ulid.ULID) (ulid.ULID, error) {
+			return b.leaser.Renew(ctx, index, leaseID)
+		},
+		func(ctx context.Context, leaseID ulid.ULID) error {
+			return b.leaser.Revoke(ctx, index, leaseID)
+		},
+		func(ctx context.Context) error {
+			// Call this function and block, renewing leases in the background
+			// until this function is done.
+			return b.compact(ctx, index)
+		},
+		25*time.Second,
```

Comment:
should this interval be configurable?
I see you added a histogram, so I assume you'd want to tweak this duration after observing how it's doing after some time?

Reply:
Yeah it definitely should be